contracts/misc/TimeLock.sol

```diff
@@ -0,0 +1,187 @@
+    mapping(uint256 => Agreement) private agreements;
+
+    uint248 public agreementCount;
+    bool public frozen;
+
+    IPoolAddressesProvider private immutable ADDRESSES_PROVIDER;
+
+    modifier onlyXToken(address asset) {
+        require(
+            msg.sender ==
+                IPool(ADDRESSES_PROVIDER.getPool()).getReserveXToken(asset),
+            Errors.CALLER_NOT_XTOKEN
+        );
+        _;
+    }
+
+    constructor(IPoolAddressesProvider provider) {
+        ADDRESSES_PROVIDER = provider;
+    }
+
+    function initialize() public initializer {
+        __Ownable_init();
+        __ReentrancyGuard_init();
+    }
+
+    function createAgreement(
+        DataTypes.AssetType assetType,
+        DataTypes.TimeLockActionType actionType,
+        address asset,
+        uint256[] calldata tokenIdsOrAmounts,
+        address beneficiary,
+        uint48 releaseTime
+    ) external onlyXToken(asset) returns (uint256) {
+        uint256 agreementId = agreementCount++;
+        agreements[agreementId] = Agreement({
+            assetType: assetType,
+            actionType: actionType,
+            asset: asset,
+            tokenIdsOrAmounts: tokenIdsOrAmounts,
+            beneficiary: beneficiary,
+            releaseTime: releaseTime,
+            isFrozen: false
+        });
+
+        emit AgreementCreated(
+            agreementId,
+            assetType,
+            actionType,
+            asset,
+            tokenIdsOrAmounts,
+            beneficiary,
+            releaseTime
+        );
+
+        return agreementId;
+    }
+
+    function _validateAndDeleteAgreement(uint256 agreementId)
+        internal
+        returns (Agreement memory)
+    {
+        Agreement memory agreement = agreements[agreementId];
+        require(msg.sender == agreement.beneficiary, "Not beneficiary");
+        require(
+            block.timestamp >= agreement.releaseTime,
+            "Release time not reached"
+        );
+        require(!agreement.isFrozen, "Agreement frozen");
+        delete agreements[agreementId];
+
+        emit AgreementClaimed(
+            agreementId,
+            agreement.assetType,
+            agreement.actionType,
+            agreement.asset,
+            agreement.tokenIdsOrAmounts,
+            agreement.beneficiary
+        );
+
+        return agreement;
+    }
+
+    function claim(uint256[] calldata agreementIds) external nonReentrant {
+        require(!frozen, "TimeLock is frozen");
+
+        for (uint256 index = 0; index < agreementIds.length; index++) {
+            Agreement memory agreement = _validateAndDeleteAgreement(
+                agreementIds[index]
+            );
+
+            if (agreement.assetType == DataTypes.AssetType.ERC20) {
+                IERC20(agreement.asset).safeTransfer(
+                    agreement.beneficiary,
+                    agreement.tokenIdsOrAmounts[0]
+                );
+            } else if (agreement.assetType == DataTypes.AssetType.ERC721) {
```

Comment:
not allow to claim moobird nft in this else logic, or the corresponding agreemenIds will be deleted